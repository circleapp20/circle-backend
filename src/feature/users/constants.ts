import { Constants as SharedConstants } from 'shared/constants';

export const Constants = {
	...SharedConstants,
	users: {
		PROFILE_IMAGE_PATH: 'images/users/profile'
	}
};
