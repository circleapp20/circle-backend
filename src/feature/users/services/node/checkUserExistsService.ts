import { runQuery } from 'core/node/database/queryRunners';
import {
	countMatchingEmailQuery,
	countMatchingUsernameQuery
} from 'feature/users/queries/countUsersQueries';

export const checkUsernameOrEmailExists = async (username: string, email: string) => {
	const totalMatchingUsernameCount = !username
		? 0
		: await runQuery(countMatchingUsernameQuery, [username]);
	const totalMatchingEmailCount = !email ? 0 : await runQuery(countMatchingEmailQuery, [email]);
	return {
		username: Boolean(totalMatchingUsernameCount),
		email: Boolean(totalMatchingEmailCount)
	};
};
