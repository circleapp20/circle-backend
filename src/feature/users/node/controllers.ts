import { Response } from 'express';
import {
	checkUsernameOrEmailExists,
	updateUserPassword,
	updateUserProfile
} from 'feature/users/node/dataService';
import { getResponseData } from 'shared/common/utilities';
import { Constants } from 'shared/constants';
import { IAuthUser, IRequest } from 'shared/types';

export const updateProfile = async (req: IRequest, res: Response) => {
	const { id }: IAuthUser = req.user;
	const data = Object.assign({}, req.body.data, { id });
	const user = await updateUserProfile(data);
	const responseData = getResponseData(user);
	res.status(Constants.status.SUCCESS).json(responseData);
};

export const searchUsernameOrEmail = async (req: IRequest, res: Response) => {
	const { username = '', email = '' }: any = req.query!;
	const exists = await checkUsernameOrEmailExists(username, email);
	const responseData = getResponseData(exists);
	res.status(Constants.status.SUCCESS).json(responseData);
};

export const resetUserPassword = async (req: IRequest, res: Response) => {
	const { id } = req.user;
	const isUpdated = await updateUserPassword(id, req.body.data.password);
	const responseData = getResponseData(isUpdated);
	res.status(Constants.status.SUCCESS).json(responseData);
};
