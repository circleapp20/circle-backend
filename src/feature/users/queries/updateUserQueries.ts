import { Users } from 'base/common/schema/users';
import { EntityManager } from 'typeorm';
import { IAddUserLocation, IUpdateUserProfile } from './updateUserTypes';

export const updateUserProfileQuery = (manager: EntityManager, values: IUpdateUserProfile) => {
	const { id, ...rest } = values;
	return manager.createQueryBuilder().update(Users).set(rest).where('id = :id', { id }).execute();
};

export const addUserLocationsQuery = (manager: EntityManager, values: IAddUserLocation) => {
	const { id, locationsId } = values;
	return manager.createQueryBuilder().relation(Users, 'locations').of(id).add(locationsId);
};
