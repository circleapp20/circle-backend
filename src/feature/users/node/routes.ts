import * as schema from 'feature/users/common/dataSchema';
import * as controller from 'feature/users/node/controllers';
import { Constants } from 'shared/constants';
import { IApiRoute } from 'shared/types';

export const usersRoutes: IApiRoute[] = [
	{
		path: '/users/profile',
		method: 'put',
		controller: controller.updateProfile,
		privileges: [Constants.privileges.USER],
		schema: schema.UpdateUserProfileSchema
	},
	{
		path: '/users/search',
		method: 'get',
		controller: controller.searchUsernameOrEmail,
		schema: schema.CheckUsernameSchema,
		type: 'query'
	},
	{
		path: '/users/reset/password',
		method: 'put',
		controller: controller.resetUserPassword,
		schema: schema.ResetPasswordSchema,
		privileges: [Constants.privileges.USER]
	}
];
