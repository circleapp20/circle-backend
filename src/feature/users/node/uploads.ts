import { Constants } from 'feature/users/constants';
import { uploadImageToFirebaseStorage } from 'shared/node/storage';

export const uploadUserProfileImage = async (id: string, image: string) => {
	if (!image || !id) return '';
	const filename = `${Constants.users.PROFILE_IMAGE_PATH}/${id}`;
	const imageURL = await uploadImageToFirebaseStorage(image, filename);
	return imageURL;
};
