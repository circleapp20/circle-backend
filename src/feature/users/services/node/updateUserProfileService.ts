import { getBadRequestError } from 'base/common/errors';
import { Constants } from 'base/constants';
import { getUserByIdQuery } from 'base/node/queries';
import bcryptjs from 'bcryptjs';
import { runInTransaction, runQuery } from 'core/node/database/queryRunners';
import { countUsersMatchingSearchQuery } from 'feature/users/queries/countUsersQueries';
import {
	addUserLocationsQuery,
	updateUserProfileQuery
} from 'feature/users/queries/updateUserQueries';
import { IAddUserLocation, IUpdateUserProfile } from 'feature/users/queries/updateUserTypes';
import { uploadUserProfileImage } from 'feature/users/uploads/node/uploadUserProfileImage';
import { EntityManager } from 'typeorm';

type UpdateUser = IUpdateUserProfile & IAddUserLocation;

export const updateUserTransaction = (data: UpdateUser) => {
	const { locationsId, id, ...rest } = data;
	const password = bcryptjs.hashSync(data.password, 12);
	return async (manager: EntityManager) => {
		const profile = Object.assign({}, rest, { password, id });
		await runQuery(updateUserProfileQuery, [profile], manager);
		// add user locations
		if (locationsId.length) {
			await runQuery(addUserLocationsQuery, [{ locationsId, id }], manager);
		}
		return runQuery(getUserByIdQuery, [data.id], manager);
	};
};

export const updateUserProfile = async (data: UpdateUser) => {
	const { username, id, password: rawPassword, image, email, phoneNumber } = data;
	if (!id) throw getBadRequestError('id is required');

	if (!rawPassword) throw getBadRequestError('password is required');

	const count = await runQuery(countUsersMatchingSearchQuery, [{ username, id }]);
	if (count !== 1) throw getBadRequestError('username already exists');

	if (phoneNumber) {
		const total = await runQuery(countUsersMatchingSearchQuery, [{ phoneNumber, id }]);
		if (total !== 1) throw getBadRequestError('phone number already exists');
	}

	if (email) {
		const total = await runQuery(countUsersMatchingSearchQuery, [{ email, id }]);
		if (total !== 1) throw getBadRequestError('email already exists');
	}

	// upload image to firebase storage bucket
	const downloadURL = await uploadUserProfileImage(id, image);

	const profile = Object.assign({}, data, { image: downloadURL });

	const user: any = await runInTransaction(updateUserTransaction(profile));
	const { password, ...other } = user;

	return other;
};

export const updateUserPassword = async (id: string, password: string) => {
	const user = await runQuery(getUserByIdQuery, [id]);
	if (!user) throw getBadRequestError('Invalid user account');

	if (bcryptjs.compareSync(password, user.password)) {
		throw getBadRequestError('Cannot enter the same password');
	}

	const hashedPassword = bcryptjs.hashSync(password, Constants.misc.BCRYPT_HASHING_SALT);

	await runQuery(updateUserProfileQuery, [{ id, password: hashedPassword }]);

	return true;
};
