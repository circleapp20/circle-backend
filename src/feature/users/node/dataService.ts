import bcryptjs from 'bcryptjs';
import {
	countMatchingEmailQuery,
	countMatchingUsernameQuery,
	countUsersMatchingSearchQuery,
	updateUserProfileQuery
} from 'feature/users/node/queries';
import { uploadUserProfileImage } from 'feature/users/node/uploads';
import { IUpdateUserProfile } from 'feature/users/types';
import { getBadRequestError } from 'shared/common/errors';
import { Constants } from 'shared/constants';
import { runInTransaction, runQuery } from 'shared/node/database';
import { getUserByIdQuery } from 'shared/node/queries';
import { EntityManager } from 'typeorm';

export const updateUserTransaction = (data: IUpdateUserProfile) => {
	return async (manager: EntityManager) => {
		const password = bcryptjs.hashSync(data.password, 12);
		const profile = Object.assign({}, data, { password });
		await runQuery(updateUserProfileQuery, [profile], manager);
		return runQuery(getUserByIdQuery, [data.id], manager);
	};
};

export const updateUserProfile = async (data: IUpdateUserProfile) => {
	const { username, id, password: rawPassword, image, email, phoneNumber } = data;
	if (!id) throw getBadRequestError('id is required');

	if (!rawPassword) throw getBadRequestError('password is required');

	const count = await runQuery(countUsersMatchingSearchQuery, [{ username, id }]);
	if (count !== 1) throw getBadRequestError('username already exists');

	if (phoneNumber) {
		const total = await runQuery(countUsersMatchingSearchQuery, [{ phoneNumber, id }]);
		if (total !== 1) throw getBadRequestError('phone number already exists');
	}

	if (email) {
		const total = await runQuery(countUsersMatchingSearchQuery, [{ email, id }]);
		if (total !== 1) throw getBadRequestError('email already exists');
	}

	// upload image to firebase storage bucket
	const downloadURL = await uploadUserProfileImage(id, image);

	const profile = Object.assign({}, data, { image: downloadURL });

	const user: any = await runInTransaction(updateUserTransaction(profile));
	const { password, ...other } = user;

	return other;
};

export const checkUsernameOrEmailExists = async (username: string, email: string) => {
	const totalMatchingUsernameCount = !username
		? 0
		: await runQuery(countMatchingUsernameQuery, [username]);
	const totalMatchingEmailCount = !email ? 0 : await runQuery(countMatchingEmailQuery, [email]);
	return {
		username: Boolean(totalMatchingUsernameCount),
		email: Boolean(totalMatchingEmailCount)
	};
};

export const updateUserPassword = async (id: string, password: string) => {
	const user = await runQuery(getUserByIdQuery, [id]);
	if (!user) throw getBadRequestError('Invalid user account');

	if (bcryptjs.compareSync(password, user.password)) {
		throw getBadRequestError('Cannot enter the same password');
	}

	const hashedPassword = bcryptjs.hashSync(password, Constants.misc.BCRYPT_HASHING_SALT);

	await runQuery(updateUserProfileQuery, [{ id, password: hashedPassword }]);

	return true;
};
