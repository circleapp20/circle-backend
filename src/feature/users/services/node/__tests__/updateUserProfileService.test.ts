import { entityManager } from 'base/testUtils/node/entityManager';
import bcryptjs from 'bcryptjs';
import { runQuery } from 'core/node/database/queryRunners';
import faker from 'faker';
import * as queries from 'feature/users/queries/updateUserQueries';
import { createUserFixture } from 'fixtures/users';
import {
	updateUserPassword,
	updateUserProfile,
	updateUserTransaction
} from '../updateUserProfileService';

jest.mock('core/node/database/queryRunners');
jest.mock('base/common/schema/users');
jest.mock('base/common/schema/fellows');
jest.mock('base/common/schema/locations');
jest.mock('bcryptjs', () => ({
	hashSync: jest.fn().mockReturnValue('$$20yy39nv93n932n92093nf92'),
	compareSync: jest.fn()
}));

beforeEach(() => jest.clearAllMocks());

describe('#updateUserTransaction', () => {
	let user: any, values: any, transaction: any;

	beforeAll(() => {
		user = createUserFixture();
		values = {
			...user,
			locationsId: Array.from(Array(2), faker.random.uuid)
		};
		transaction = updateUserTransaction(values);
	});

	test('should update a user profile', async () => {
		const updateMock = jest.spyOn(queries, 'updateUserProfileQuery');
		await transaction(entityManager);
		expect(updateMock).not.toHaveBeenCalledWith(entityManager, values);
	});

	test('should add locations to user', async () => {
		const spy = jest.spyOn(queries, 'addUserLocationsQuery');
		await transaction(entityManager);
		expect(spy).toHaveBeenCalledWith(entityManager, {
			locationsId: values.locationsId,
			id: user.id
		});
	});

	test("should not add locations to user if there aren't any", async () => {
		const spy = jest.spyOn(queries, 'addUserLocationsQuery');
		const runTransaction = updateUserTransaction(
			Object.assign({}, values, { locationsId: [] })
		);
		await runTransaction(entityManager);
		expect(spy).not.toHaveBeenCalled();
	});
});

describe('#updateUserProfile', () => {
	let user: any, values: any;

	beforeAll(() => {
		user = createUserFixture();
		values = {
			...user,
			locationsId: Array.from(Array(2), faker.random.uuid)
		};
	});

	test('should update profile with valid details', async () => {
		entityManager.getCount.mockReturnValueOnce(1).mockReturnValueOnce(1).mockReturnValueOnce(1);
		entityManager.execute.mockReturnValueOnce(true);
		entityManager.getOne.mockReturnValueOnce(values);
		const user = await updateUserProfile(values);
		expect(user).toBeDefined();
	});

	test('should throw error if username already exists', () => {
		(runQuery.mockResolvedValueOnce as any)(2);
		expect(updateUserProfile(values)).rejects.toThrowErrorMatchingSnapshot();
	});

	test('should throw error if id is invalid', () => {
		const data = Object.assign({}, values, { id: '' });
		expect(updateUserProfile(data)).rejects.toThrowErrorMatchingSnapshot();
	});

	test('should throw error if password is invalid', () => {
		const data = Object.assign({}, values, { password: '' });
		expect(updateUserProfile(data)).rejects.toThrowErrorMatchingSnapshot();
	});

	test('should throw error if phone number already exists', () => {
		entityManager.getCount.mockReturnValueOnce(1).mockReturnValueOnce(2);
		const data = Object.assign({}, values, { email: '' });
		expect(updateUserProfile(data)).rejects.toThrowErrorMatchingSnapshot();
	});

	test('should throw error if another user has the same email', () => {
		entityManager.getCount.mockReturnValueOnce(1).mockReturnValueOnce(2);
		const data = Object.assign({}, values, { phoneNumber: '' });
		expect(updateUserProfile(data)).rejects.toThrowErrorMatchingSnapshot();
	});
});

describe('#updateUserPassword', () => {
	beforeEach(() => {
		(<any>runQuery).mockResolvedValue({ id: 'xi93m', password: '2f3n92' });
		(bcryptjs as any).compareSync.mockReturnValue(false);
	});

	test('should throw error if user does not exists', (done) => {
		(<any>runQuery).mockResolvedValueOnce(null);
		updateUserPassword('xi93xm', 'testing').catch((error) => {
			expect(error.message).toBe('Invalid user account');
			done();
		});
	});

	test('should throw error if password matches old password', (done) => {
		(bcryptjs as any).compareSync.mockReturnValueOnce(true);
		updateUserPassword('xi93m', 'testing').catch((error) => {
			expect(error.message).toBe('Cannot enter the same password');
			done();
		});
	});

	test('should hash the new password', async () => {
		await updateUserPassword('xi93m', 'testing');
		expect(bcryptjs.hashSync).toHaveBeenCalledWith(
			expect.stringMatching('testing'),
			expect.any(Number)
		);
	});

	test('should update user password', async () => {
		await updateUserPassword('xi93m', 'testing');
		expect(runQuery).toHaveBeenLastCalledWith(
			expect.any(Function),
			expect.arrayContaining([
				expect.objectContaining({
					id: expect.any(String),
					password: expect.any(String)
				})
			])
		);
	});
});
