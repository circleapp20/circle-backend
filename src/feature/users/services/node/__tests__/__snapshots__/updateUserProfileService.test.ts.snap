// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`#updateUserProfile should throw error if another user has the same email 1`] = `"email already exists"`;

exports[`#updateUserProfile should throw error if id is invalid 1`] = `"id is required"`;

exports[`#updateUserProfile should throw error if password is invalid 1`] = `"password is required"`;

exports[`#updateUserProfile should throw error if phone number already exists 1`] = `"phone number already exists"`;

exports[`#updateUserProfile should throw error if username already exists 1`] = `"username already exists"`;
